### PR TITLE
feat: Add support for ParquetSource in Dagger capable of parsing primitive data types

### PR DESCRIPTION
**Pull request by Meghajit**
PR for https://github.com/odpf/dagger/issues/99

### CODE REVIEW

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    private final SimpleGroupDeserializer simpleGroupDeserializer;
+    private long currentRecordIndex;
+    private final ParquetFileReader reader;
+    private long rowCount;
+    private RecordReader<Group> recordReader;
+    private final MessageType schema;
+    private static final Logger LOGGER = LoggerFactory.getLogger(PrimitiveReader.class.getName());
+
+    public PrimitiveReader(String filePath, SimpleGroupDeserializer simpleGroupDeserializer) throws IOException {
+        this.hadoopFilePath = new Path(filePath);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java:39]: will need some discussion to implement it alternatively

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        this.simpleGroupDeserializer = simpleGroupDeserializer;
+        Path path = new Path(filePath);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java:41]: Will do

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/TimestampProtoHandler.java [hunk 3]
```diff
@@ -93,10 +98,34 @@ public Object transformFromPostProcessor(Object field) {
...
+            throw new IllegalArgumentException(errMessage);
+        }
+        return transformFromSimpleGroup((SimpleGroup) field);
+    }
+
+    /* conversion from ms to nanos borrowed from Instant.java class and inlined here for performance reasons */
+    private Object transformFromSimpleGroup(SimpleGroup simpleGroup) {
+        String fieldName = fieldDescriptor.getName();
+        if (SimpleGroupValidation.checkFieldExistsAndIsInitialized(simpleGroup, fieldName)) {
+            long timeInMillis = simpleGroup.getLong(fieldName, 0);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/TimestampProtoHandler.java:119]: will need some discussion how/when to implement it

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandler.java [hunk 4]
```diff
@@ -30,9 +32,24 @@ public boolean canHandle() {
 
     @Override
     public Object getValue(Object field) {
+        if (field instanceof SimpleGroup) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandler.java:35]: will do

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandler.java [hunk 5]
```diff
@@ -30,9 +32,24 @@ public boolean canHandle() {
 
     @Override
     public Object getValue(Object field) {
+        if (field instanceof SimpleGroup) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandler.java:35]: will do

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/ByteStringPrimitiveTypeHandler.java [hunk 6]
```diff
@@ -32,9 +34,24 @@ public boolean canHandle() {
 
     @Override
     public Object getValue(Object field) {
+        if (field instanceof SimpleGroup) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/ByteStringPrimitiveTypeHandler.java:37]: will do

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandler.java [hunk 7]
```diff
@@ -30,9 +32,24 @@ public boolean canHandle() {
 
     @Override
     public Object getValue(Object field) {
+        if (field instanceof SimpleGroup) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandler.java:35]: will do

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java [hunk 8]
```diff
@@ -47,11 +48,25 @@ public static Row createRow(DynamicMessage proto, int extraColumns) {
...
+        List<FieldDescriptor> descriptorFields = descriptor.getFields();
+        Row row = new Row(descriptorFields.size() + extraColumns);
+        for (FieldDescriptor fieldDescriptor : descriptorFields) {
+            ProtoHandler protoHandler = ProtoHandlerFactory.getProtoHandler(fieldDescriptor);
+            row.setField(fieldDescriptor.getIndex(), protoHandler.transformFromParquet(simpleGroup));
+        }
+        return row;
+    }
+
+    public static Row createRow(Descriptors.Descriptor descriptor, SimpleGroup simpleGroup) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java:66]: It will be used in MessageProtoHandler and RepeatedMessageProtoHandler as part of #100 
Can remove it now if you want.

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandlerTest.java [hunk 9]
```diff
@@ -4,17 +4,26 @@
...
 import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.MockitoAnnotations.initMocks;
 
 public class BooleanPrimitiveTypeHandlerTest {
+    @Before
+    public void setup() {
+        initMocks(this);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandlerTest.java:25]: will remove

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java [hunk 10]
```diff
@@ -90,4 +93,50 @@ public void shouldReturnEmptyArrayOnNull() {
...
     }
 
+    @Test
+    public void shouldFetchParsedValueForFieldOfTypeDoubleInSimpleGroup() {
+        Descriptors.FieldDescriptor fieldDescriptor = TestBookingLogMessage.getDescriptor().findFieldByName("cash_amount");
+        GroupType parquetSchema = org.apache.parquet.schema.Types.requiredGroup()
+                .required(DOUBLE).named("cash_amount")
+                .named("TestGroupType");
+        SimpleGroup simpleGroup = new SimpleGroup(parquetSchema);
+        simpleGroup.add("cash_amount", Double.MAX_VALUE);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java:103]: Sure, will use a normal value.
I'm not aware of such a practice though. What's the reason for it ?

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandlerTest.java [hunk 11]
```diff
@@ -92,4 +96,49 @@ public void shouldReturnEmptyArrayOnNull() {
...
     }
 
+    @Test
+    public void shouldFetchParsedValueForFieldOfTypeFloatInSimpleGroup() {
+        Descriptors.FieldDescriptor fieldDescriptor = TestBookingLogMessage.getDescriptor().findFieldByName("amount_paid_by_cash");
+        GroupType parquetSchema = requiredGroup()
+                .required(FLOAT).named("amount_paid_by_cash")
+                .named("TestGroupType");
+        SimpleGroup simpleGroup = new SimpleGroup(parquetSchema);
+        simpleGroup.add("amount_paid_by_cash", Float.MIN_VALUE);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandlerTest.java:106]: Sure, will use a normal value.
I'm not aware of such a practice though. What's the reason for it ?

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java [hunk 12]
```diff
@@ -94,4 +97,50 @@ public void shouldReturnEmptyArrayOnNull() {
...
 
+    @Test
+    public void shouldFetchParsedValueForFieldOfTypeIntegerInSimpleGroup() {
+        Descriptors.FieldDescriptor fieldDescriptor = TestBookingLogMessage.getDescriptor().findFieldByName("cancel_reason_id");
+
+        GroupType parquetSchema = org.apache.parquet.schema.Types.requiredGroup()
+                .required(INT32).named("cancel_reason_id")
+                .named("TestGroupType");
+        SimpleGroup simpleGroup = new SimpleGroup(parquetSchema);
+        simpleGroup.add("cancel_reason_id", Integer.MIN_VALUE);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java:108]: Sure, will use a normal value.
I'm not aware of such a practice though. What's the reason for it ?

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java [hunk 13]
```diff
@@ -92,4 +95,49 @@ public void shouldReturnEmptyArrayOnNull() {
...
     }
 
+    @Test
+    public void shouldFetchParsedValueForFieldOfTypeLongInSimpleGroup() {
+        Descriptors.FieldDescriptor fieldDescriptor = TestAggregatedSupplyMessage.getDescriptor().findFieldByName("s2_id");
+        GroupType parquetSchema = org.apache.parquet.schema.Types.requiredGroup()
+                .required(INT64).named("s2_id")
+                .named("TestGroupType");
+        SimpleGroup simpleGroup = new SimpleGroup(parquetSchema);
+        simpleGroup.add("s2_id", Long.MIN_VALUE);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java:105]: Sure, will use a normal value.
I'm not aware of such a practice though. What's the reason for it ?

File: dagger-core/src/main/java/io/odpf/dagger/core/source/SourceName.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+import static io.odpf.dagger.core.utils.Constants.STREAM_SOURCE_DETAILS_SOURCE_NAME_KAFKA;
+import static io.odpf.dagger.core.utils.Constants.STREAM_SOURCE_DETAILS_SOURCE_NAME_PARQUET;
+
+public enum SourceName {
+    @SerializedName(STREAM_SOURCE_DETAILS_SOURCE_NAME_KAFKA)
+    KAFKA,
+    @SerializedName(STREAM_SOURCE_DETAILS_SOURCE_NAME_PARQUET)
+    PARQUET,
+    @SerializedName(STREAM_SOURCE_DETAILS_SOURCE_NAME_KAFKA_CONSUMER)
+    KAFKA_CONSUMER
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/SourceName.java:15]: KAFKA & KAFKA_CONSUMER was suggested by @prakharmathur82 . What are your views ? :-)

The problem with using names like `KAFKA & OLD_KAFKA` is that if the current KafkaSource also becomes legacy, what do we name the latest one ?

`KAFKA_SOURCE & KAFKA_CONSUMER` sounds good to me. But in that case, can we also rename `PARQUET` to `PARQUET_SOURCE` ?

File: dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        private final Configuration configuration;
+        private final StencilClientOrchestrator stencilClientOrchestrator;
+
+        public Builder(StreamConfig streamConfig, Configuration configuration, StencilClientOrchestrator stencilClientOrchestrator) {
+            this.streamConfig = streamConfig;
+            this.configuration = configuration;
+            this.stencilClientOrchestrator = stencilClientOrchestrator;
+        }
+
+        public Builder(StreamConfig streamConfig, Configuration configuration) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java:41]: I think is remnants of the previous code when we were creating different stream types like Json, Proto, Parquet, etc. Json did not require stencil , hence this constructor was used.

I'll remove it.

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.parquet.io.MessageColumnIO;
+import org.apache.parquet.io.RecordReader;
+import org.apache.parquet.schema.MessageType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+public class PrimitiveReader implements FileRecordFormat.Reader<Row> {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java:26]: The current reader is the most basic(aka primitive) implementation of a parquet reader which is operating at a very low level and uses RowGroups, Pages and Records ( internal abstractions of a Parquet File) to extract data out. There is no high level library used on top of these abstractions, except the ParquetFileReader library to read the file. Furthermore, it considers an entire ParquetFile as one split and cannot break it down further for more parallelism. That's why it is the most primitive reader

My thinking is that, as the codebase matures, we might design/use better parquet readers which operate at a more high level by maybe extending this reader.

File: dagger-core/src/test/java/io/odpf/dagger/core/source/StreamsFactoryTest.java [hunk 17]
```diff
@@ -33,89 +25,33 @@ public class StreamsFactoryTest {
...
-        List<Stream> streams = StreamsFactory.getStreams(configuration, stencilClientOrchestrator, telemetryExporter);
-
-        Assert.assertEquals(DataTypes.JSON, streams.get(0).getInputDataType());
-    }
-
-    @Test
-    public void shouldCreateMultipleStreams() {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/StreamsFactoryTest.java:85]: It's still there, just under a different name 
` public void shouldReturnListOfStreamTypesCreatedFromConfiguration()`

File: dagger-core/src/test/java/io/odpf/dagger/core/source/StreamTypeTest.java [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+        when(streamConfig.getEventTimestampFieldIndex()).thenReturn("5");
+        when(streamConfig.getDataType()).thenReturn("PROTO");
+        when(streamConfig.getProtoClass()).thenReturn("com.tests.TestMessage");
+        when(streamConfig.getSchemaTable()).thenReturn("data_stream");
+        when(stencilClientOrchestrator.getStencilClient()).thenReturn(stencilClient);
+        when(stencilClient.get("com.tests.TestMessage")).thenReturn(TestBookingLogMessage.getDescriptor());
+        StreamType.Builder builder = new StreamType.Builder(streamConfig, configuration, stencilClientOrchestrator);
+        StreamType streamType = builder.build();
+
+        assertTrue(streamType.getDaggerSource() instanceof ParquetDaggerSource);
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/StreamTypeTest.java:122]: will try to do this

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java:103]: when using extreme values it is very easy to run into issues where because of framework/serde those values might be set in case of invalid values, which will be missed if we use this in our tests (happened once with us, not in this repo). Using a more specific value avoid such corner cases

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandlerTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandlerTest.java:106]: explained above

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java:108]: explained above

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java:105]: explained above

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java:26]: primitive looks more like primitive data reader and can be confusing.

[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java:26]: Do you want to propose some alternative names ?

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java:66]: > Even in those let's use 0 explicitly
Umm, but why though ? Even for Proto above, we are using a similar constructor

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java:103]: Sure

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandler.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/10f20eedf67bc808d157382def84321d3fb007d9

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandler.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/FloatPrimitiveTypeHandler.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/10f20eedf67bc808d157382def84321d3fb007d9

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/ByteStringPrimitiveTypeHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/ByteStringPrimitiveTypeHandler.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/10f20eedf67bc808d157382def84321d3fb007d9

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandler.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandler.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/10f20eedf67bc808d157382def84321d3fb007d9

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/2dfb01be3e238b79472d1e674d3bf0ba2ae40ca4

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java]: Fixed as part of review comment https://github.com/odpf/dagger/pull/121/commits/18ee55d76473e71dfba2f5409242cc93d2daa275

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandlerTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/BooleanPrimitiveTypeHandlerTest.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/814e4dd42bd848f6a6614b8c8af8a782063381c0

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/LongPrimitiveTypeHandlerTest.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/26b77f6d1960d59cb7ced3d5a9751b3c7ecedfa0

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/IntegerPrimitiveTypeHandlerTest.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/26961c11522d0239e94d5fab15db7ed8e0a21e67

File: dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/test/java/io/odpf/dagger/common/serde/proto/protohandler/typehandler/DoublePrimitiveTypeHandlerTest.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/f106599978047dbfe5e9f30eeb4c5ecec77dbd7c

File: dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/01bacb84a4a1a9e86cc68f624438b46bc6ba48d7

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/TimestampProtoHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/TimestampProtoHandler.java:111]: As discussed, for timestamps of type seconds+nanos in the simple group, they can be handled as part of #100.

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/proto/protohandler/RowFactory.java:66]: As discussed, keeping this,  as this will be needed in the next issue #100 when nested simple groups would need to be converted to Row without any extra columns.

File: dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+import io.odpf.dagger.core.deserializer.DaggerDeserializerFactory;
+import lombok.Getter;
+import org.apache.flink.api.common.eventtime.WatermarkStrategy;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.types.Row;
+
+import java.io.Serializable;
+
+public class StreamType implements Serializable {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/StreamType.java]: As discussed, renamed StreamType back to Stream so as to maintain symmetry with the config `STREAMS`
Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/4a82762d532bb68243a4910287179693c4751d76

File: dagger-core/src/main/java/io/odpf/dagger/core/source/SourceName.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/SourceName.java:15]: As discussed, renamed `KAFKA` to `KAFKA_SOURCE` and PARQUET to `PARQUET_SOURCE`.
Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/fcac8d4bff9a6fdbf67e37d81b7f901dd827b6d0

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/PrimitiveReader.java]: As discussed, renamed `PrimitiveReader` to `ParquetReader`.
Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/a867b9e6cfdbedfde0a0236e7fe7183696ea1921

File: dagger-core/src/test/java/io/odpf/dagger/core/source/StreamTypeTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/test/java/io/odpf/dagger/core/source/StreamTypeTest.java]: Fixed as part of commit https://github.com/odpf/dagger/pull/121/commits/3d29d9ac44c29e6cebdd442fb7382c5cf7ecffcf

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    private RecordReader<Group> recordReader;
+    private final MessageType schema;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ParquetReader.class.getName());
+
+    private ParquetReader(Path hadoopFilePath, SimpleGroupDeserializer simpleGroupDeserializer, ParquetFileReader parquetFileReader) throws IOException {
+        this.hadoopFilePath = hadoopFilePath;
+        this.simpleGroupDeserializer = simpleGroupDeserializer;
+        this.parquetFileReader = parquetFileReader;
+        this.schema = this.parquetFileReader.getFileMetaData().getSchema();
+        changeReaderPosition(this.parquetFileReader.readNextRowGroup());
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java:40]: We are initialising the reader position here so that when Flink calls `read` internally in the next step, it reads from the beginning of the file. 
I don't think this will be computationally expensive. Additionally, this is initialization code which should be in the constructor I think.

[the remark this answers is not in the document]
Comment (gauravsinghania) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java:40]: It will be faster than the network. However, still it is reading all the chunks of the file which will be slower. Would prefer it be initialized lazily.

[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java]: @gauravsinghania Refactored as part of commit https://github.com/odpf/dagger/pull/121/commits/0c25068fb7a44fdf1d46d548ddbc6ef81e3f7701

File: dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+            changeReaderPosition(page);
+        }
+        return readAndDeserialize();
+    }
+
+    @Nullable
+    @Override
+    public Row read() throws IOException {
+        if (!isRecordReaderInitialized) {
+            return initializeRecordReader();
```
Comment (gauravsinghania) [inline dagger-core/src/main/java/io/odpf/dagger/core/source/parquet/reader/ParquetReader.java:80]: at present, initialize is doing multiple things. Initializing as well as reading. Can we break it down and rather just use readRecordsFromPage after initialization is done? 
so that way, you only initialize when it is not initialized and the rest of the flow is the same irrespective of flow (first time or later).

**Comment by lavkesh** [comment]
🎉 This is a big one :) Kudos @Meghajit
